### PR TITLE
Adding TxSuccessful to PayeAtsService. 

### PR DESCRIPTION
**Pull request by chrisjameswright**
Requested for service monitoring dashboard, bringing PAYE service in line with SA service

### CODE REVIEW

File: app/services/PayeAtsService.scala [hunk 1]
```diff
@@ -28,15 +31,20 @@ import scala.concurrent.Future
...
   def middleConnector: MiddleConnector
+  def auditService: AuditService
 
-  def getPayeATSData(nino: Nino, taxYear: Int)(implicit hc: HeaderCarrier):Future[Either[HttpResponse,PayeAtsData]] = {
+  def getPayeATSData(nino: Nino, taxYear: Int)(implicit hc: HeaderCarrier, request: PayeAuthenticatedRequest[_]):Future[Either[HttpResponse,PayeAtsData]] = {
      middleConnector.connectToPayeATS(nino,taxYear) map { response =>
        response status match {
-         case OK =>  Right(response.json.as[PayeAtsData])
+         case OK =>  {
+           sendAuditEvent(taxYear)
```
Comment (oc-ste) [inline app/services/PayeAtsService.scala:45]: What happens if sendAuditEvent fails?

File: app/controllers/auth/PayeAuthAction.scala [hunk 2]
```diff
@@ -43,10 +44,11 @@ class PayeAuthActionImpl @Inject()(override val authConnector: AuthConnector,
         HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
 
       authorised(ConfidenceLevel.L200 and AuthNino(hasNino = true) and CredentialStrength(CredentialStrength.strong))
-        .retrieve(Retrievals.nino) {
-          case Some(nino) => {
+        .retrieve(Retrievals.externalId and Retrievals.nino) {
+          case Some(userId) ~ Some(nino) => {
             block {
               PayeAuthenticatedRequest(
+                userId,
```
Comment (oc-ste) [inline app/controllers/auth/PayeAuthAction.scala:51]: Don't have the ticket but can we check why we need this external ID?

File: app/services/PayeAtsService.scala [hunk 3]
```diff
@@ -48,4 +56,16 @@ trait PayeAtsService {
...
   }
+
+  private def sendAuditEvent(taxYear: Int)(implicit hc: HeaderCarrier, request: PayeAuthenticatedRequest[_]) = {
+    auditService.sendEvent(
+      auditType = AuditTypes.Tx_SUCCEEDED,
+      details = Map(
+        "userId" -> request.userId,
+        "userNino" -> request.nino.nino,
+        "taxYear" -> taxYear.toString,
+        "time" -> new Date().toString
```
Comment (oc-ste) [inline app/services/PayeAtsService.scala:67]: The time stamp is stored when it logs anyway, do we need this field?

File: app/services/PayeAtsService.scala [hunk 4]
```diff
@@ -48,4 +56,16 @@ trait PayeAtsService {
        }
      }
   }
+
+  private def sendAuditEvent(taxYear: Int)(implicit hc: HeaderCarrier, request: PayeAuthenticatedRequest[_]) = {
+    auditService.sendEvent(
+      auditType = AuditTypes.Tx_SUCCEEDED,
+      details = Map(
+        "userId" -> request.userId,
+        "userNino" -> request.nino.nino,
```
Comment (oc-ste) [inline app/services/PayeAtsService.scala:65]: This nino could not be the same nino being passed in to `middleConnector.connectToPayeATS(nino`

File: app/controllers/auth/PayeAuthAction.scala (hunk 2 above)
Comment (shikkhasingh) [reply app/controllers/auth/PayeAuthAction.scala:51]: I don't think we need to audit externalId . Looked into the splunk and couldn't find out any advantage of having this field as we are already logging NINO that is unique to the user and we are going to have sessionIDs as well in audit logs.

File: app/services/PayeAtsService.scala (hunk 3 above)
Comment (shikkhasingh) [reply app/services/PayeAtsService.scala:67]: Yeah completely agree Timestamp is by default there in the logs so we don't need this time field.
Going to remove it.

File: app/services/PayeAtsService.scala (hunk 4 above)
Comment (shikkhasingh) [reply app/services/PayeAtsService.scala:65]: We are passing request.nino from controller that is being passed to the connector.
So both nino are same but i understand the confusion will update the code accordingly.

File: app/services/PayeAtsService.scala (hunk 1 above)
Comment (shikkhasingh) [reply app/services/PayeAtsService.scala:45]: As discussed going to log the exception and prevent user from using the service if  sendAuditEvent fails.
Need to confirm from performance analyst regarding the cache bit.

Comment (chrisjameswright) [reply app/services/PayeAtsService.scala:45]: > Also do we want to audit every time, even when we get data from cache?
This is an outstanding question, which will need to be clarified as part of the ticket. I've emailed about getting this created properly but haven't heard back yet.

File: app/controllers/auth/PayeAuthAction.scala (hunk 2 above)
Comment (chrisjameswright) [reply app/controllers/auth/PayeAuthAction.scala:51]: This was requested by the analyst building the dashboard, as it makes it equivalent to the payload for the SA part of the service.

File: app/services/PayeAtsService.scala (hunk 3 above)
Comment (chrisjameswright) [reply app/services/PayeAtsService.scala:67]: This was requested by the analyst building the dashboard, as it makes it equivalent to the payload for the SA part of the service, but I agree in this case it should be clarified.

File: app/services/PayeAtsService.scala (hunk 4 above)
Comment (chrisjameswright) [reply app/services/PayeAtsService.scala:65]: I agree their is a risk that this isn't the NINO we used (even though the call to the backend should fail if they don't match). I would agree with changing and using the one from the call.

**Comment by chrisjameswright** [comment]
Please keep in mind that this shouldn't be merged until it has been cleared with TXM (who, understandably, are very busy with urgent work currently).

**Pull request merged**